### PR TITLE
[CALCITE] Extraction of procedure schema

### PR DESCRIPTION
**Pull request by AndrewPochapsky**
Added extraction of create procedure signature and hooked up call statements to the function validation infrastructure.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java [hunk 1]
```diff
@@ -138,6 +139,31 @@ public CalciteCatalogReader withSchemaPath(List<String> schemaPath) {
...
 
+  /**
+   * Gets the procedure with the given name if it exists.
+   *
+   * @param names The name of the procedure, list represents possibility of a
+   *              compound identifier
+   *
+   * @return The found procedure, or null
+   */
+  private Procedure getProcedure(List<String> names) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:150]: This looks identical to `getMacro()`. Can we condense these two cases into a single function?

File: core/src/main/java/org/apache/calcite/schema/impl/Procedure.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Schema class for all user defined procedures.
+ */
+public class Procedure implements Function {
+  private final List<FunctionParameter> parameters;
+
+  /**
+   * Creates a {@code Procedure}.
+   *
+   * @param parameters The procedure parameters
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/schema/impl/Procedure.java:34]: nit: Must not be null

File: core/src/main/java/org/apache/calcite/sql/SqlFunctionCategory.java [hunk 3]
```diff
@@ -65,8 +67,8 @@ public boolean isUserDefined() {
     return properties.contains(USER_DEFINED);
   }
 
-  public boolean isUserDefinedFunction() {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlFunctionCategory.java:68]: Why are we deleting this existing function? Is it not used anymore?

File: core/src/main/java/org/apache/calcite/sql/SqlCreateProcedure.java [hunk 4]
```diff
@@ -105,6 +105,7 @@ public SqlCreateProcedure(SqlParserPos pos, SqlCreateSpecifier createSpecifier,
 
   @Override public void validate(SqlValidator validator,
       SqlValidatorScope scope) {
+    validator.addProcedureToSchema(this);
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlCreateProcedure.java:108]: Why not follow the pattern used for other nodes that call back to SqlValidatorImpl with something like `validate<NodeType>()`? In this case, that would be `validateCreateProcedure()`. That function could take care of adding the procedure to the schema and validating the underlying scripting statement.

File: parsing/dialects/intermediate/dialects/dialect1/src/test/java/org/apache/calcite/test/Dialect1ValidatorTest.java [hunk 5]
```diff
@@ -240,6 +240,50 @@ public SqlNode parseAndValidate(String sql) {
...
+        + "return 1";
+    String query = "CALL foo(1)";
+    String query2 = "SELECT foo(1)";
+    sql(procedureDdl).ok();
+    sql(functionDdl).ok();
+    sql(query).type("ANY NOT NULL");
+    sql(query2).type("RecordType(INTEGER NOT NULL EXPR$0) NOT NULL");
+  }
+
+  @Test public void testCreateProcedureSchemaInvalidNumberFails() {
```
Comment (dasch-1) [inline parsing/dialects/intermediate/dialects/dialect1/src/test/java/org/apache/calcite/test/Dialect1ValidatorTest.java:272]: `testCreateProcedureSchemaInvalidNumberOfParamsFails`

File: core/src/main/java/org/apache/calcite/sql/SqlFunctionCategory.java (hunk 3 above)
Comment (AndrewPochapsky) [reply core/src/main/java/org/apache/calcite/sql/SqlFunctionCategory.java:68]: Yeah, I was initially using it in the previous feature but forgot to remove it.

File: core/src/main/java/org/apache/calcite/sql/SqlCreateProcedure.java (hunk 4 above)
Comment (AndrewPochapsky) [reply core/src/main/java/org/apache/calcite/sql/SqlCreateProcedure.java:108]: Sure, that makes more sense.

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java [hunk 6]
```diff
@@ -156,9 +171,17 @@ private Procedure getProcedure(List<String> names) {
...
+      Function function = null;
+      switch (category) {
+      case USER_DEFINED_PROCEDURE:
+        function = schema.getProcedure(name, caseSensitive);
+        break;
+      case USER_DEFINED_MACRO:
+        function = schema.getMacro(name, caseSensitive);
+        break;
+      }
+      if (function != null) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:183]: nit: Can we remove this and just `return function` unconditionally at the end of this function?

Comment (AndrewPochapsky) [reply core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:183]: I'm not 100% sure on this one. All tests pass if I return function at the end but I don't know if that will work for all cases. Could it happen that function is not null and then later becomes null via more iterations of the loop?

Comment (dasch-1) [reply core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:183]: Ah I see what you mean. In that case, let's just leave it as is.

**Pull request merged**